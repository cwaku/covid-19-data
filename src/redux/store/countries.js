import moment from 'moment';
import { createSlice } from '@reduxjs/toolkit';
import axios from 'axios';
// set axios default url
axios.defaults.baseURL = 'https://api.covid19tracking.narrativa.com/api/';

const date = moment(new Date()).format('YYYY-MM-DD');

const countriesSlice = createSlice({
  name: 'countries',
  initialState: {
    countries: [],
    selectedCountry: '',
    searchCountry: '',
  },
  reducers: {
    getCountriesSuccess: (state, action) => ({ ...state, countries: action.payload }),
    getSelectedCountrySuccess: (state, action) => ({ ...state, selectedCountry: action.payload }),
    getSearchedCountrySuccess: (state, action) => ({ ...state, searchCountry: action.payload }),
    getCountriesFailure: (state, action) => ({ ...state, countries: action.payload }),
  },
});

export default countriesSlice.reducer;

// Actions

export const {
  getCountriesSuccess,
  getSelectedCountrySuccess,
  getSearchedCountrySuccess,
  getCountriesFailure,
} = countriesSlice.actions;

export const fetchCountries = () => async (dispatch) => {
  try {
    const response = await axios.get(date);
    if (response) {
      const { dates, total } = response.data;
      const { countries } = dates[date];
      const countriesList = Object.values(countries);

      const singleCountry = countriesList.map((country) => (
        {
          name: country.name,
          confirmedCases: country.today_confirmed,
          deaths: country.today_deaths,
          recovered: country.today_recovered,
          active: country.today_open_cases,
          id: country.id,
        }
      ));

      const totalConfirmedCases = total.today_confirmed;
      const allData = { singleCountry, totalConfirmedCases };
      dispatch(getCountriesSuccess(allData));
    }
  } catch (error) {
    dispatch(getCountriesFailure(error));
  }
};
