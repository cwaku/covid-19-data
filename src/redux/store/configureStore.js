import { configureStore } from '@reduxjs/toolkit'
import thunk from 'redux-thunk'
import logger from 'redux-logger'
import { combineReducers } from 'redux'
import countries from './countries'

const reducer = combineReducers({
  // reducers go here
  countries,
})

const store = configureStore({
  reducer,
  middleware: [thunk, logger],
})

export default store;