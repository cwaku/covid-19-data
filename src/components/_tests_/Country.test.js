import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import Countries from '../homepage/Country';
import store from '../../redux/store/configureStore';
import { fetchCountries } from '../../redux/store/countries';

describe('Test reducers', () => {
  let state;
  beforeEach(() => {
    state = store.getState();
  });

  it('should return the initial state', () => {
    expect(state.countries.countries).toEqual([]);
  });

  it('should return the correct state after fetching countries', () => {
    store.dispatch(fetchCountries());
    state = store.getState();
    expect(state.countries.countries).toEqual([]);
  });
});

describe('Test Countries component', () => {
  it('should render the Countries component correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Countries />
        </BrowserRouter>
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
