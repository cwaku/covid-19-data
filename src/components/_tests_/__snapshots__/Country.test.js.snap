// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Countries component should render the Countries component correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-8atqhb"
  >
    <h2
      class="stats"
    >
      STATS BY COUNTRY:
    </h2>
    <div
      class="MuiGrid-root MuiGrid-container css-10mzr36-MuiGrid-root"
    />
  </div>
</div>
`;
